File: scripts/serve.mjs
import { createServer } from 'node:http';
import { createReadStream } from 'node:fs';
import { stat } from 'node:fs/promises';
import { resolve, extname, sep } from 'node:path';
const root = resolve(import.meta.dirname, '..');
const types = { '.html': 'text/html; charset=utf-8', '.js': 'text/javascript', '.css': 'text/css', '.json': 'application/json', '.wasm': 'application/wasm', '.ttf': 'font/ttf', '.svg': 'image/svg+xml', '.txt': 'text/plain' };
createServer(async (req, res) => {
  try {
    const path = decodeURIComponent(new URL(req.url, 'http://localhost').pathname);
    if (path === '/recite' || path === '/recite/') { res.writeHead(302, { Location: '/' }).end(); return; }
    const rel = path === '/' ? 'index.html' : path.replace(/^\//, '');
    let file = resolve(root, rel);
    if (!file.startsWith(root + sep) || rel.split('/').some(part => part.startsWith('.'))) { res.writeHead(403).end(); return; }
    let info = await stat(file).catch(() => null);
    if (!info?.isFile()) { file = resolve(root, 'public', rel); info = await stat(file).catch(() => null); }
    if (!info?.isFile()) { res.writeHead(404).end('Not found'); return; }
    res.writeHead(200, { 'Content-Type': types[extname(file)] || 'application/octet-stream', 'Content-Length': info.size, 'Cache-Control': 'no-cache' });
    if (req.method === 'HEAD') res.end(); else createReadStream(file).pipe(res);
  } catch { res.writeHead(400).end('Bad request'); }
}).listen(5173, '127.0.0.1', () => console.log('HafizAssist: http://localhost:5173'));

File: scripts/serve.py
"""Serve the source app and public assets locally using Python 3 only."""
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import unquote, urlsplit

ROOT = Path(__file__).resolve().parent.parent


class AppHandler(SimpleHTTPRequestHandler):
    extensions_map = {
        **SimpleHTTPRequestHandler.extensions_map,
        '.js': 'text/javascript; charset=utf-8',
        '.wasm': 'application/wasm',
        '.json': 'application/json; charset=utf-8',
        '.ttf': 'font/ttf',
        '.onnx': 'application/octet-stream',
    }

    def send_head(self):
        path = unquote(urlsplit(self.path).path)
        if path in ('/recite', '/recite/'):
            self.send_response(302)
            self.send_header('Location', '/')
            self.send_header('Content-Length', '0')
            self.end_headers()
            return None
        relative = {'/': 'index.html', '/mobile': 'mobile.html'}.get(path, path.lstrip('/'))
        if '\\' in relative or any(part.startswith('.') for part in relative.split('/')):
            self.send_error(403)
            return None
        try:
            for base in (ROOT, ROOT / 'public'):
                file = (base / relative).resolve()
                if base not in file.parents:
                    self.send_error(403)
                    return None
                if not file.is_file():
                    continue
                stream = file.open('rb')
                self.send_response(200)
                self.send_header('Content-Type', self.guess_type(str(file)))
                self.send_header('Content-Length', str(file.stat().st_size))
                self.send_header('Cache-Control', 'no-cache')
                self.end_headers()
                return stream
            self.send_error(404, 'File not found')
        except (OSError, ValueError):
            self.send_error(404, 'File unavailable')
        return None


if __name__ == '__main__':
    with ThreadingHTTPServer(('127.0.0.1', 5173), AppHandler) as server:
        print('Recite Al Quran: http://localhost:5173', flush=True)
        print('Mobile page: http://localhost:5173/mobile.html', flush=True)
        print('Press Ctrl+C to stop.', flush=True)
        try:
            server.serve_forever()
        except KeyboardInterrupt:
            pass

